In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
#Importing the dataset
dataset =pd.read_csv('50_Startups.csv')


In [3]:
print(dataset)

    R&D Spend  Administration  Marketing Spend       State     Profit
0   165349.20       136897.80        471784.10    New York  192261.83
1   162597.70       151377.59        443898.53  California  191792.06
2   153441.51       101145.55        407934.54     Florida  191050.39
3   144372.41       118671.85        383199.62    New York  182901.99
4   142107.34        91391.77        366168.42     Florida  166187.94
5   131876.90        99814.71        362861.36    New York  156991.12
6   134615.46       147198.87        127716.82  California  156122.51
7   130298.13       145530.06        323876.68     Florida  155752.60
8   120542.52       148718.95        311613.29    New York  152211.77
9   123334.88       108679.17        304981.62  California  149759.96
10  101913.08       110594.11        229160.95     Florida  146121.95
11  100671.96        91790.61        249744.55  California  144259.40
12   93863.75       127320.38        249839.44     Florida  141585.52
13   91992.39       

In [4]:
X = dataset.iloc[:,:-1]
Y = dataset.iloc[:,4]

In [5]:
print(X)

    R&D Spend  Administration  Marketing Spend       State
0   165349.20       136897.80        471784.10    New York
1   162597.70       151377.59        443898.53  California
2   153441.51       101145.55        407934.54     Florida
3   144372.41       118671.85        383199.62    New York
4   142107.34        91391.77        366168.42     Florida
5   131876.90        99814.71        362861.36    New York
6   134615.46       147198.87        127716.82  California
7   130298.13       145530.06        323876.68     Florida
8   120542.52       148718.95        311613.29    New York
9   123334.88       108679.17        304981.62  California
10  101913.08       110594.11        229160.95     Florida
11  100671.96        91790.61        249744.55  California
12   93863.75       127320.38        249839.44     Florida
13   91992.39       135495.07        252664.93  California
14  119943.24       156547.42        256512.92     Florida
15  114523.61       122616.84        261776.23    New Yo

In [6]:
print(Y)

0     192261.83
1     191792.06
2     191050.39
3     182901.99
4     166187.94
5     156991.12
6     156122.51
7     155752.60
8     152211.77
9     149759.96
10    146121.95
11    144259.40
12    141585.52
13    134307.35
14    132602.65
15    129917.04
16    126992.93
17    125370.37
18    124266.90
19    122776.86
20    118474.03
21    111313.02
22    110352.25
23    108733.99
24    108552.04
25    107404.34
26    105733.54
27    105008.31
28    103282.38
29    101004.64
30     99937.59
31     97483.56
32     97427.84
33     96778.92
34     96712.80
35     96479.51
36     90708.19
37     89949.14
38     81229.06
39     81005.76
40     78239.91
41     77798.83
42     71498.49
43     69758.98
44     65200.33
45     64926.08
46     49490.75
47     42559.73
48     35673.41
49     14681.40
Name: Profit, dtype: float64


In [7]:
#convert the column into categorical columns(convert it to one hot encoding)
states = pd.get_dummies(X['State'],drop_first=True)
print(states)

    Florida  New York
0         0         1
1         0         0
2         1         0
3         0         1
4         1         0
5         0         1
6         0         0
7         1         0
8         0         1
9         0         0
10        1         0
11        0         0
12        1         0
13        0         0
14        1         0
15        0         1
16        0         0
17        0         1
18        1         0
19        0         1
20        0         0
21        0         1
22        1         0
23        1         0
24        0         1
25        0         0
26        1         0
27        0         1
28        1         0
29        0         1
30        1         0
31        0         1
32        0         0
33        1         0
34        0         0
35        0         1
36        1         0
37        0         0
38        0         1
39        0         0
40        0         0
41        1         0
42        0         0
43        0         1
44        

In [8]:
#Drop the state column
X=X.drop('State',axis=1)


In [9]:
print(X)

    R&D Spend  Administration  Marketing Spend
0   165349.20       136897.80        471784.10
1   162597.70       151377.59        443898.53
2   153441.51       101145.55        407934.54
3   144372.41       118671.85        383199.62
4   142107.34        91391.77        366168.42
5   131876.90        99814.71        362861.36
6   134615.46       147198.87        127716.82
7   130298.13       145530.06        323876.68
8   120542.52       148718.95        311613.29
9   123334.88       108679.17        304981.62
10  101913.08       110594.11        229160.95
11  100671.96        91790.61        249744.55
12   93863.75       127320.38        249839.44
13   91992.39       135495.07        252664.93
14  119943.24       156547.42        256512.92
15  114523.61       122616.84        261776.23
16   78013.11       121597.55        264346.06
17   94657.16       145077.58        282574.31
18   91749.16       114175.79        294919.57
19   86419.70       153514.11             0.00
20   76253.86

In [10]:
#concat the dummy variables
X = pd.concat([X,states],axis=1)

In [11]:
print(X)

    R&D Spend  Administration  Marketing Spend  Florida  New York
0   165349.20       136897.80        471784.10        0         1
1   162597.70       151377.59        443898.53        0         0
2   153441.51       101145.55        407934.54        1         0
3   144372.41       118671.85        383199.62        0         1
4   142107.34        91391.77        366168.42        1         0
5   131876.90        99814.71        362861.36        0         1
6   134615.46       147198.87        127716.82        0         0
7   130298.13       145530.06        323876.68        1         0
8   120542.52       148718.95        311613.29        0         1
9   123334.88       108679.17        304981.62        0         0
10  101913.08       110594.11        229160.95        1         0
11  100671.96        91790.61        249744.55        0         0
12   93863.75       127320.38        249839.44        1         0
13   91992.39       135495.07        252664.93        0         0
14  119943

In [12]:
#Splitting the dataset into Training set and Test set
from sklearn.model_selection import train_test_split
X_train,X_test,Y_train,Y_test = train_test_split(X,Y,test_size=0.2,random_state=0)

In [13]:
print(X_train)

    R&D Spend  Administration  Marketing Spend  Florida  New York
33   55493.95       103057.49        214634.81        1         0
35   46014.02        85047.44        205517.64        0         1
26   75328.87       144135.98        134050.07        1         0
34   46426.07       157693.92        210797.67        0         0
18   91749.16       114175.79        294919.57        1         0
7   130298.13       145530.06        323876.68        1         0
14  119943.24       156547.42        256512.92        1         0
45    1000.23       124153.04          1903.93        0         1
48     542.05        51743.15             0.00        0         1
29   65605.48       153032.06        107138.38        0         1
15  114523.61       122616.84        261776.23        0         1
30   61994.48       115641.28         91131.24        1         0
32   63408.86       129219.61         46085.25        0         0
16   78013.11       121597.55        264346.06        0         0
42   23640

In [14]:
print(X_test)


    R&D Spend  Administration  Marketing Spend  Florida  New York
28   66051.52       182645.56        118148.20        1         0
11  100671.96        91790.61        249744.55        0         0
10  101913.08       110594.11        229160.95        1         0
41   27892.92        84710.77        164470.71        1         0
2   153441.51       101145.55        407934.54        1         0
27   72107.60       127864.55        353183.81        0         1
38   20229.59        65947.93        185265.10        0         1
31   61136.38       152701.92         88218.23        0         1
22   73994.56       122782.75        303319.26        1         0
4   142107.34        91391.77        366168.42        1         0


In [15]:
print(Y_train)


33     96778.92
35     96479.51
26    105733.54
34     96712.80
18    124266.90
7     155752.60
14    132602.65
45     64926.08
48     35673.41
29    101004.64
15    129917.04
30     99937.59
32     97427.84
16    126992.93
42     71498.49
20    118474.03
43     69758.98
8     152211.77
13    134307.35
25    107404.34
5     156991.12
17    125370.37
40     78239.91
49     14681.40
1     191792.06
12    141585.52
37     89949.14
24    108552.04
6     156122.51
23    108733.99
36     90708.19
21    111313.02
19    122776.86
9     149759.96
39     81005.76
46     49490.75
3     182901.99
0     192261.83
47     42559.73
44     65200.33
Name: Profit, dtype: float64


In [16]:
print(Y_test)

28    103282.38
11    144259.40
10    146121.95
41     77798.83
2     191050.39
27    105008.31
38     81229.06
31     97483.56
22    110352.25
4     166187.94
Name: Profit, dtype: float64


In [17]:
#Fitting Multiple Linear Regression to Training set
from sklearn.linear_model import LinearRegression
reg = LinearRegression()
reg.fit(X_train,Y_train);

In [18]:
#Predicting the Test set results
Y_pred = reg.predict(X_test)

In [19]:
print(Y_pred)

[103015.20159796 132582.27760816 132447.73845174  71976.09851258
 178537.48221055 116161.24230165  67851.69209676  98791.73374687
 113969.43533012 167921.0656955 ]


In [20]:
from sklearn.metrics import r2_score
score = r2_score(Y_test,Y_pred)

In [21]:
print(score)

0.9347068473282423


#### The R square value comes out to be 0.9347 which is a good result and hence we can say it is a good model.